tests/test_target.py

```diff
@@ -14,17 +14,31 @@
 # limitations under the License.
 #
 
+"""Module for testing targets."""
+
 import os
 import shutil
 import tempfile
 from unittest import TestCase
 
 from devlib import LocalLinuxTarget
+from devlib.utils.misc import load_struct_from_yaml
 
 
 class TestReadTreeValues(TestCase):
+    """Class testing Target.read_tree_values_flat()"""
+
+    TARGETS_CONFIG_FILE = os.path.join(os.path.dirname(os.path.realpath(__file__)),
```

Comment:
If the intention is to be able to perform these tests on different device types based on the config, do we want the parsing / initialization to be performed within each test case? 
Should these actions be performed in some setup method instead that can be reused?

Reply:
The final product is something like this: https://github.com/ARM-software/devlib/pull/667/commits/a7d0b9d80bc47c22e88a398d5a0bcb6c0c332b2b
Yes, there are different target types and there can be multiple entries for a target.
I read config file first and see which targets need to be tested. So, I could keep target classes as is.

Reply:
Hmm.. I see. Apologies in advanced if I'm missing the obvious but in the final product do we end up with this code just for a single `test_read_multiline_values` unittest? 
Does this mean this will need to be duplicated when we add another test or test Class?

Also does this require setting up a new connection to the target for each test? I'm wondering what the overhead/performance would be on this?

Reply:
There is only one `test_read_multiline_values()` function. It parses the YAML file and executes the test specified in `run_test()` for each target in the YAML file. YAML file has necessary connection information and QEMU params for each target. Depending on the target type, running `run_test()` can take ~1 (LocalLinux target) or ~15 secs (for QEMU target).

Final version of `TestReadTreeValues()` looks like this:
```Python
class TestReadTreeValues(TestCase):
    def test_read_multiline_values(self):
        def run_test(target):
            ...

        target_configs = load_struct_from_yaml(self.TARGETS_CONFIG_FILE)
        ...

        if target_configs.get('AndroidTarget') is not None:
            ...
                target = AndroidTarget(
                    connection_settings=entry['connection_settings'],
                    conn_cls=lambda **kwargs: AdbConnection(adb_as_root=True, **kwargs),
                )
                run_test(target)

        if target_configs.get('LinuxTarget') is not None:
            ...
                target = LinuxTarget(connection_settings=entry['connection_settings'])
                run_test(target)

        if target_configs.get('LocalLinuxTarget') is not None:
            ...
                target = LocalLinuxTarget(connection_settings=entry['connection_settings'])
                run_test(target)

        if target_configs.get('QEMUTargetRunner') is not None:
           ...
                with QEMUTargetRunner(
                        qemu_params=qemu_params,
                        connection_settings=connection_settings,
                ) as qemu_target:
                    run_test(qemu_target.target)
```
And this is a sample YAML file:
```Yaml
AndroidTarget:
  entry-0:
    connection_settings:
      device: 'emulator-5554'
  entry-1:
    connection_settings:
      device: 'emulator-5556'

LinuxTarget:
  entry-0:
    connection_settings:
      host: 'example.com'
      username: 'username'
      password: 'password'

LocalLinuxTarget:
  entry-0:
    connection_settings:
      unrooted: True

QEMUTargetRunner:
  entry-0:
    qemu_params:
      kernel_image: '/devlib/tools/buildroot/buildroot-v2023.11.1-aarch64/output/images/Image'

  entry-1:
    qemu_params:
      kernel_image: '/devlib/tools/buildroot/buildroot-v2023.11.1-x86_64/output/images/bzImage'
      arch: 'x86_64'
      cmdline: 'console=ttyS0'
```

Reply:
I see, thanks for the information.
 
Although we have only have a single test case `test_read_multiline_values` at the moment, if we have this nice infrastructure to be able to run these tests and have more meaningful output, I'm thinking if we were expecting to add more test cases going forward, how would that fit into the current implementation? For example if we wanted to add a `test_read_escape_values` test case that checks certain characters are escaped correctly.

The parsing / creation of the various targets looks like it should be mostly generic and not tied to this single test case, therefore I'm wondering if there is a common location this would be better suited to be shared?

Reply:
That totally makes sense to me.

How about the new version? If it looks OK, I'm gonna update other PRs.
Moving to `pytest` is the simplest solution I could find to prevent duplicate target config reading, passing target parameter to the test function (as well as future ones).

Thanks.